### ISSUE TITLE
Embedding space collapsing after training using cosine similarity loss

### ISSUE THREAD

**Issue by rishabh16196** [UKPLab/sentence-transformers#3469]
Hi, 
I am trying to finetune mini-lm-l6-v2 model using trainer and using cosine similarity loss. I have a dataset of 30M annotated pairs containing both positive and negative cases(fairly equally distributed) in my training dataset. After training, the model is working awesomely well to correctly give high cosine similarity to positive pairs(in my dataset all positives are between 0.6 -1 and negatives are labeled as 0) . 
But the problem is with anything which is not in the training dataset, ie two random sentences are giving 99.6 cosine similarity. Almost any random text is outputting this similarity. >99%.
On further analysis, I saw that many texts the individual dimensions are very close. of very random texts. 

Can anyone help with what might be the issue?

**Comment by johneckberg** [comment]
Hi @rishabh16196!

While cosine similarity loss does a good job of pushing positive pairs together and the specified negative pairs apart, it fails to distribute embeddings uniformly across the entire embedding space.

Compare this with something like [Multiple Negative Ranking](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/losses/MultipleNegativesRankingLoss.py) loss. MNR loss makes the assumption that if two sentences in the training data aren't positively paired, that we can use them as a negative pair. This process of pushing a bunch of vector pairs apart is what helps uniformly distribute the vectors across the embedding space. 

That all being said, I haven't seen an embedding space collapse this severely before, though I haven't trained very many models using cosine similarity loss. Have you tried using [CoSENT loss](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/losses/CoSENTLoss.py) for this model as well? CoSENT loss is recommended over cosine similarity loss. Although I'm not confident it would help with this issue of a collapsed embedding space, it could be worth experimenting.

**Comment by rishabh16196** [comment]
Hi @johneckberg 
Yes when experimenting with MNRL and CoSENT loss, the embeddings are distributed uniformly and I am not getting random pairs giving this high of a similarity.
But for my use case I want to calibrate the scores as well. With cosent loss, I am losing control over the final scores for positive pairs. I have a 35M dataset with annotated scores and would like my positives to follow the scores that are there in dataset. 
What do you suggest for calibrating scores post training?

Attaching a visualisation of tensorboard projector for the trained model for reference of what I mean by embedding space collapsing.

Unrelated texts getting clustered together :
<img width="1370" height="959" alt="Image" src="https://github.com/user-attachments/assets/4ac8e1e8-2fa5-4ab1-9631-cc60ff3986c3" />

Good examples : 

<img width="1373" height="980" alt="Image" src="https://github.com/user-attachments/assets/31141df0-f7f6-451c-a2d8-c11f2e27a3ff" />

**Comment by johneckberg** [comment]
@rishabh16196 
Unfortunately I'm not currently aware of any methods for calibrating the similarity scores of text/image embedding models. I have been poking around the past few days, but have not been able to find anything on the topic. It might be worth trying to combine CoSENT and Cosine Similarity loss in order to try and get the benefits of both? That is to say, a new loss that calculates the both losses and adds them together on the forward pass or in the `compute_loss_from_embeddings` method. I think this would be relatively straightforward as they share the same input format.

**Comment by ddofer** [comment]
Suggestion (I have a similar issue, albeit not solved): Use Euclidean loss. That'll help avoid losing features relating to the scale/magnitude of individual vectors/features
